### ISSUE TITLE
💬 RFC: Catalog plugin: `spec.type` attribute on all entity kinds

### ISSUE THREAD

**Issue by jvanalstyne-mdsol** [backstage/backstage#24978]
### 🔖 Need

Our company has its own internal terminology for the various levels/entities in our product map, and we want to reflect this as best as we can in our backstage catalog UI. 

With the introduction of subdomains, our backstage catalog is closer than ever to perfectly matching our product map. However, now that there are multiple levels of Domains, we are faced with the difficulty of trying to differentiate between certain entity types that fall into the Domain category. For example, there are some instances where we want one Domain to be labeled as a "Product Family" vs. its subdomain as a "Product," but other Domains where different terminology is needed.

We also have use cases for differentiating between certain Systems and Users based on our internal terminology.

We already solved this issue for Group, Component, and other entity kinds using their existing `spec.type` attribute. It seems logical to enable this attribute for all entity kinds to make the catalog more flexible. Selfishly, having this as a standard attribute would also make implementing custom UI easier :)

### 🎉 Proposal

I would like to suggest adding an _optional_ `spec.type` attribute on the "Domain", "System", and "User" entity kinds. (These are the only catalog entity kinds that do not have this attribute available.) Having it be optional would make it possible for anyone who doesn't need the information to just ignore it. No clear impacts to privacy or security.

### 〽️ Alternatives

Using metadata/annotations - discarded because:
1. Precedent exists for using `spec.type` on other entity kinds for this same purpose
2. Having different attributes for the same kind of information introduces complexity in UI and other plugins that depend on the catalog plugin

### ❌ Risks

I don't know if this is actually the case, but I imagine this could potentially break any logic that depends on these entity kinds to **not** have the`spec.type` attribute.

In general, though, I am not sure if this change could somehow be incompatible with anything important, or if I'm just missing something in general. This is where I would most appreciate any input 🙏 

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by freben** [comment]
Hm I'm torn. On the one hand for hierarchical things, like the Domain now is, there's more clear need to explain what type each layer is. For systems, and _in particular_ users, it starts getting much muddier. As in, we could add the field, but we need to also provide guidance on _how_ it is to be used; what its semantics are. For a user, I think there will be several competing possible interpretations of how it's to be used. Contract type (`employee`, `contractor`, `intern`, ...), title (`engineering manager`, `cto`, ...), discipline (`backend developer`, `designer`, ...), seniority (`junior`, `senior 2`, ...), primary purpose in squad (`manager`, `member`), etc etc. And the thing is, almost all of those should ideally instead have been metadata _on the relation_ to the group, not on the user itself, because its value could be different in relation to different groups.

So in summary - maybe adding it for Domain could be a viable core model addition. Mmmmmmaybe for System, after some pondering. Not for User, that's just too early to know what it'd even mean.

And either way - it's perfectly fine to have private extensions to the model! The majority of adopters do, I'd say, and there's nothing wrong with that. The main reason to want it open sourced would be if open source plugins wanted to ascribe meaning to them and start writing plugins whose logic depend on the field. But here - I'm not sure that's even the case.

**Comment by jvanalstyne-mdsol** [comment]
Thanks for the thoughtful response @freben!!

After reading your logic, I think I agree that the User kind probably doesn't need a `spec.type` attribute.

And I can see how needs could also differ for the System kind based on organizational needs and circumstances. My team is indeed working on a plugin we hope to open source at some point, so perhaps we could reconsider adding `spec.type` to the System kind if it would even be necessary.

That said, as you suggest, I agree that being able to differentiate each layer of Domain/Subdomain is a fairly clear cut use. I think it could have immediate benefit for organizations incorporating subdomains into their catalog. Would it be appropriate for me to move forward with making a PR to add the type field to the Domain model?

**Comment by freben** [comment]
Yeah, you could go ahead for Domain, thank you. Maybe System too, honestly. They will have to be optional though, as `type?: string;` in the typescript sense, so as not to break current usage out there.

**Comment by jvanalstyne-mdsol** [comment]
Will do. Thanks again!
